File: docs/frontend/encryptionHelper.ts
/**
 * RSA-OAEP Encryption Helper for JavaScript/TypeScript
 *
 * Dependencies:
 *   npm install node-forge
 *   npm install --save-dev @types/node-forge  (for TypeScript)
 *
 * This helper fetches the RSA public key from the backend and encrypts
 * each credential field individually using RSA-OAEP with SHA-256 padding.
 * The JSON request shape stays the same — only the values are encrypted.
 */

import * as forge from "node-forge";

interface EncryptionHelperConfig {
    /** Base URL of the backend API (e.g., "https://api.example.com") */
    baseUrl: string;
}

class EncryptionHelper {
    private baseUrl: string;
    private publicKey: forge.pki.rsa.PublicKey | null = null;

    constructor(config: EncryptionHelperConfig) {
        this.baseUrl = config.baseUrl;
    }

    /** Fetch the RSA public key from the backend. */
    private async ensureKeyLoaded(): Promise<forge.pki.rsa.PublicKey> {
        if (this.publicKey) return this.publicKey;

        const response = await fetch(`${this.baseUrl}/api/v1/auth/public-key`);
        if (!response.ok) {
            throw new Error(`Failed to fetch public key: ${response.status}`);
        }

        const data = await response.json();
        this.publicKey = forge.pki.publicKeyFromPem(data.public_key);
        return this.publicKey;
    }

    /**
     * Encrypt a single string field using RSA-OAEP/SHA-256.
     * @returns Base64-encoded ciphertext.
     */
    async encryptField(plaintext: string): Promise<string> {
        const publicKey = await this.ensureKeyLoaded();

        const encrypted = publicKey.encrypt(plaintext, "RSA-OAEP", {
            md: forge.md.sha256.create(),
            mgf1: { md: forge.md.sha256.create() },
        });

        return forge.util.encode64(encrypted);
    }

    /**
     * Encrypt login credentials. Returns an object with the same shape
     * as a plain login request, but with encrypted values.
     *
     * @returns `{ username: "<encrypted>", password: "<encrypted>" }`
     */
    async encryptLoginBody(
        username: string,
        password: string
    ): Promise<{ username: string; password: string }> {
        return {
            username: await this.encryptField(username),
            password: await this.encryptField(password),
        };
    }

    /**
     * Encrypt a new password for the password-reset endpoint.
     * @returns Base64-encoded RSA-OAEP encrypted string.
     */
    async encryptNewPassword(newPassword: string): Promise<string> {
        return this.encryptField(newPassword);
    }
}

export { EncryptionHelper, EncryptionHelperConfig };

// ---------------------------------------------------------------------------
// Usage Examples
// ---------------------------------------------------------------------------
//
// const helper = new EncryptionHelper({
//   baseUrl: "https://your-api-domain.com",
// });
//
// // --- Login ---
// async function login(username: string, password: string) {
//   const body = await helper.encryptLoginBody(username, password);
//   // body = { username: "<encrypted>", password: "<encrypted>" }
//
//   const response = await fetch(
//     "https://your-api-domain.com/api/v1/auth/login",
//     {
//       method: "POST",
//       headers: { "Content-Type": "application/json" },
//       body: JSON.stringify(body),
//     }
//   );
//
//   return response.json();
// }
//
// // --- Password Reset ---
// async function resetPassword(
//   userId: number,
//   otp: string,
//   newPassword: string
// ) {
//   const encryptedPassword = await helper.encryptNewPassword(newPassword);
//
//   const response = await fetch(
//     "https://your-api-domain.com/api/v1/auth/password-reset/verify-otp",
//     {
//       method: "POST",
//       headers: { "Content-Type": "application/json" },
//       body: JSON.stringify({
//         user_id: userId,
//         otp: otp,
//         new_password: encryptedPassword,
//       }),
//     }
//   );
//
//   return response.json();
// }

File: docs/frontend/encryption_helper.dart
/// RSA-OAEP Encryption Helper for Flutter/Dart
///
/// Dependencies (add to pubspec.yaml):
///   encrypt: ^5.0.3
///   pointycastle: ^3.9.1
///   http: ^1.2.0
///
/// This helper fetches the RSA public key from the backend and encrypts
/// each credential field individually using RSA-OAEP with SHA-256 padding.

import 'dart:convert';
import 'package:encrypt/encrypt.dart';
import 'package:pointycastle/asymmetric/api.dart';
import 'package:http/http.dart' as http;

class EncryptionHelper {
  /// Base URL of the backend API (configure per environment)
  final String baseUrl;

  /// Cached RSA public key
  RSAPublicKey? _publicKey;

  /// Cached Encrypter instance
  Encrypter? _encrypter;

  EncryptionHelper({required this.baseUrl});

  /// Fetch the RSA public key from the backend.
  Future<void> _ensureKeyLoaded() async {
    if (_encrypter != null) return;

    final response = await http.get(
      Uri.parse('$baseUrl/api/v1/auth/public-key'),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to fetch public key: ${response.statusCode}');
    }

    final body = jsonDecode(response.body);
    final pemString = body['public_key'] as String;

    // Parse the PEM public key
    final parser = RSAKeyParser();
    _publicKey = parser.parse(pemString) as RSAPublicKey;
    _encrypter = Encrypter(RSA(
      publicKey: _publicKey,
      encoding: RSAEncoding.OAEP,
      digest: RSADigest.SHA256,
    ));
  }

  /// Encrypt a single string field using RSA-OAEP/SHA-256.
  /// Returns a base64-encoded ciphertext string.
  Future<String> encryptField(String plaintext) async {
    await _ensureKeyLoaded();
    final encrypted = _encrypter!.encrypt(plaintext);
    return encrypted.base64;
  }

  /// Encrypt login credentials and return the request body.
  ///
  /// Returns a Map with `username` and `password` fields, each individually
  /// RSA-OAEP encrypted (base64). The JSON shape is identical to a plain
  /// login request — only the values are encrypted.
  Future<Map<String, String>> encryptLoginBody({
    required String username,
    required String password,
  }) async {
    return {
      'username': await encryptField(username),
      'password': await encryptField(password),
    };
  }

  /// Encrypt a new password for the password-reset endpoint.
  ///
  /// Returns the RSA-OAEP encrypted (base64) new_password string.
  Future<String> encryptNewPassword(String newPassword) async {
    return await encryptField(newPassword);
  }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
//
// final encryptionHelper = EncryptionHelper(
//   baseUrl: 'https://your-api-domain.com',
// );
//
// // --- Login ---
// final loginBody = await encryptionHelper.encryptLoginBody(
//   username: 'john_doe',
//   password: 'my_secure_password',
// );
//
// final loginResponse = await http.post(
//   Uri.parse('https://your-api-domain.com/api/v1/auth/login'),
//   headers: {'Content-Type': 'application/json'},
//   body: jsonEncode(loginBody),
//   // Sends: {"username": "<encrypted>", "password": "<encrypted>"}
// );
//
// // --- Password Reset ---
// final encryptedPassword = await encryptionHelper.encryptNewPassword(
//   'new_secure_password',
// );
//
// final resetResponse = await http.post(
//   Uri.parse('https://your-api-domain.com/api/v1/auth/password-reset/verify-otp'),
//   headers: {'Content-Type': 'application/json'},
//   body: jsonEncode({
//     'user_id': 123,
//     'otp': '456789',
//     'new_password': encryptedPassword,
//   }),
// );
